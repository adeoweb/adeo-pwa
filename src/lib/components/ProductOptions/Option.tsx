import React, { FunctionComponent, useMemo } from 'react';
import { useOption } from '@magento/peregrine/lib/talons/ProductOptions/useOption';
import { TConfigurableProductOption } from 'src/lib/types/graphql/Product';
import getItemKey from 'src/lib/components/ProductOptions/utils/getItemKey';
import getListComponent from 'src/lib/components/ProductOptions/utils/getListComponent';

type TProductOptionProps = {
    option: TConfigurableProductOption;
    onSelectionChange: (optionId: string, selection: number) => void;
    selectedValue?: string;
};

const Option: FunctionComponent<TProductOptionProps> = ({
    option: {
        attribute_code: attributeCode,
        attribute_id: attributeId,
        label,
        values
    },
    onSelectionChange,
    selectedValue
}) => {
    const { handleSelectionChange, initialSelection } = useOption({
        attribute_id: attributeId,
        label,
        onSelectionChange,
        selectedValue,
        values
    });

    const ValueList = useMemo(() => getListComponent(attributeCode), [
        attributeCode
    ]);

    if (!attributeCode || !attributeId || !label || !values) {
        return null;
    }

    return (
        <div className="product-single-filter">
            <label>{label}</label>
            <ValueList
                getItemKey={getItemKey}
                selectedValue={initialSelection}
                items={values}
                onSelectionChange={handleSelectionChange}
            />
        </div>
    );
};

export default Option;
