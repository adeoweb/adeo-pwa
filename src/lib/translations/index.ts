import lt from './lt/translation';
import en from './en/translation';

export default {
    en,
    lt
};
