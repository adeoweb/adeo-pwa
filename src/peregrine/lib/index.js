// components
export { default as PeregrineContextProvider } from './PeregrineContextProvider';
