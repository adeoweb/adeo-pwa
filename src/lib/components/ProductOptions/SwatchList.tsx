import React, { FunctionComponent, useMemo } from 'react';
import Swatch from './Swatch';
import { TValueListProps } from 'src/lib/components/ProductOptions/ProductOptionsTypes';

const SwatchList: FunctionComponent<TValueListProps> = ({
    getItemKey,
    selectedValue,
    items,
    onSelectionChange
}) => {
    const swatches = useMemo(
        () =>
            items.map(item => {
                const isSelected =
                    Boolean(selectedValue.label) &&
                    item.label === selectedValue.label;

                return (
                    <Swatch
                        key={getItemKey(item)}
                        isSelected={isSelected}
                        item={item}
                        onClick={onSelectionChange}
                    />
                );
            }),
        [getItemKey, selectedValue.label, items, onSelectionChange]
    );

    return <ul className="config-swatch-list">{swatches}</ul>;
};

SwatchList.displayName = 'SwatchList';

export default SwatchList;
