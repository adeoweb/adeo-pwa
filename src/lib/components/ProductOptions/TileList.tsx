import React, { FunctionComponent, useMemo } from 'react';
import Tile from 'src/lib/components/ProductOptions/Tile';
import { TValueListProps } from 'src/lib/components/ProductOptions/ProductOptionsTypes';

const TileList: FunctionComponent<TValueListProps> = ({
    getItemKey,
    selectedValue,
    items,
    onSelectionChange
}) => {
    const tiles = useMemo(
        () =>
            items.map(item => {
                const isSelected =
                    Boolean(selectedValue.label) &&
                    item.label === selectedValue.label;

                return (
                    <Tile
                        key={getItemKey(item)}
                        isSelected={isSelected}
                        item={item}
                        onClick={onSelectionChange}
                    />
                );
            }),
        [getItemKey, selectedValue.label, items, onSelectionChange]
    );

    return <ul className="config-size-list">{tiles}</ul>;
};

TileList.displayName = 'TileList';

export default TileList;
