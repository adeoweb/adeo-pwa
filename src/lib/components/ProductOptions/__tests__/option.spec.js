import React from 'react';
import { act } from 'react-test-renderer';
import { createTestInstance } from '@magento/peregrine/lib';

import Option from '../Option';
import SwatchList from 'src/lib/components/ProductOptions/SwatchList';
import TileList from 'src/lib/components/ProductOptions/TileList';

jest.mock('src/lib/components/ProductOptions/SwatchList', () => () => <i />);
jest.mock('src/lib/components/ProductOptions/TileList', () => () => <i />);

const onSelectionChangeMock = jest.fn();
const option = {
    attribute_id: '1',
    attribute_code: 'foo',
    label: 'Foo',
    values: [
        {
            store_label: 'red',
            value_index: 0
        },
        {
            store_label: 'blue',
            value_index: 1
        }
    ]
};

test('renders Option component correctly', () => {
    const component = createTestInstance(
        <Option option={option} onSelectionChange={onSelectionChangeMock} />
    );

    expect(component.toJSON()).toMatchSnapshot();
});

test('renders a SwatchList for color attributes', () => {
    const { root } = createTestInstance(
        <Option
            option={{
                ...option,
                attribute_code: 'color',
                label: 'Color'
            }}
            onSelectionChange={onSelectionChangeMock}
        />
    );

    expect(root.findByType(SwatchList)).toBeTruthy();
});

test('renders a Tile List for other attributes', () => {
    const { root } = createTestInstance(
        <Option option={option} onSelectionChange={onSelectionChangeMock} />
    );

    expect(root.findByType(TileList)).toBeTruthy();
});

test('calls onSelectionChange callback on selection change', () => {
    const { root } = createTestInstance(
        <Option option={option} onSelectionChange={onSelectionChangeMock} />
    );
    const { onSelectionChange } = root.findByType(TileList).props;
    const nextSelection = 1;

    act(() => {
        onSelectionChange(nextSelection);
    });

    expect(onSelectionChangeMock).toHaveBeenCalledTimes(1);
    expect(onSelectionChangeMock).toHaveBeenNthCalledWith(
        1,
        '1',
        nextSelection
    );
});
