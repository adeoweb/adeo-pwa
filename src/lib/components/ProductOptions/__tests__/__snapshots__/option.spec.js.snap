// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Option component correctly 1`] = `
<div
  className="product-single-filter"
>
  <label>
    Foo
  </label>
  <i />
</div>
`;
